src/CodeGen_PTX_Dev.cpp

```diff
@@ -791,6 +810,325 @@ bool CodeGen_PTX_Dev::supports_atomic_add(const Type &t) const {
 
 }  // namespace
 
+struct MatrixMultiplyInfo {
+    Expr A;
+    Expr B;
+    Expr C;
+
+    bool is_valid() const {
+        return A.defined() && B.defined() && C.defined();
+    }
+};
+
+MatrixMultiplyInfo is_matrix_multiply(const For *loop) {
+    if ((loop->for_type == ForType::Serial) && is_const_zero(loop->min) && is_const(loop->extent)) {
+        // The body of the loop can be a LetStmt or a Store
+        const LetStmt *let = loop->body.as<LetStmt>();
+        const Store *store = loop->body.as<Store>();
+
+        // If it is a let and the body is a store, use it
+        if (let && let->body.as<Store>()) {
+            store = let->body.as<Store>();
+        }
+
+        if (store) {
+            //  matmul$1[t26] = (float32)matmul$1[t26]
+            //      + (float32((float16)A[((A.stride.1*t34) - t33) + matmul$1.s1.k$x])
+            //      * *float32((float16)B[(B.stride.1*matmul$1.s1.k$x) + t35]))
+            const Expr wild_f32x = Variable::make(Float(32), "*");
+            const Expr wild_f16x = Variable::make(Float(16), "*");
+            const Expr acc_pattern = wild_f32x + f32(wild_f16x) * f32(wild_f16x);
+            vector<Expr> matches;
+            if (expr_match(acc_pattern, store->value, matches)) {
+                // (float32)matmul$1[t26]
+                const Load *load_c = matches[0].as<Load>();
+                // (float16) A[((A.stride.1 * t34) - t33) + matmul$1.s1.k$x]
+                const Load *load_a = matches[1].as<Load>();
+                // (float16)B[(B.stride.1*matmul$1.s1.k$x) + t35]
+                const Load *load_b = matches[2].as<Load>();
+
+                if (load_c && load_a && load_b) {
+                    // Check if the load is loading from the same place where the store is storing
+                    // i.e. if this is an update operation
+                    if (load_c->name == store->name && equal(load_c->index, store->index)) {
+                        // Yes, this is an update operation, now let's check cast to see if it is a matmul that can be converted
+                        // to wmma intrinsic
+                        const Expr wild_i32x = Variable::make(Int(32), "*");
+                        // Check if the reduction domain is being used in both A and B to
+                        // validate the matrix multiply
+                        const Expr load_a_pattern = wild_i32x + wild_i32x;
+                        const Expr load_b_pattern = wild_i32x * wild_i32x + wild_i32x;
+
+                        vector<Expr> matches_a, matches_b;
+                        const bool match_a = expr_match(load_a_pattern, load_a->index, matches_a);
+                        const bool match_b = expr_match(load_b_pattern, load_b->index, matches_b);
+                        if (match_a && match_b) {
```

Comment:
Should this also try commuting load_a_pattern with load_b_pattern? It seems OK to have them in either order.

Reply:
That is a fair point. I improved the code in c031f495b4af8870aafc97c1137ade51c61ff4d0